### PR TITLE
WIP: Ability to stream output from http endpoints

### LINKED ISSUE #14288: The ability to stream chunked responses from an http endpoint
### How to reproduce

1. Create a kubernetes cluster (`kind create cluster`)
2. Expose the API using `kubectl proxy --port=8080`
3. Run `http get --raw --full --max-time 30s http://127.0.0.1:8080/api/v1/namespaces/default/pods?watch=true`

### Describe the solution you'd like

Something like `http get --raw --full --max-time 30s --stream  http://127.0.0.1:8080/api/v1/namespaces/default/pods?watch=true` to keep outputting until command is terminated, timed out or the response completes.

### Describe alternatives you've considered

_No response_

### Additional context and details

_No response_

### PR DESCRIPTION
**Pull request by alex-kattathra-johnson**
fixes #14288 

# Description

Adding an optional `--stream` flag to stream output from `http get` command via websocket.

# User-Facing Changes

New optional flag called `--stream`

# Tests + Formatting

Added new tests to test the new flag and to ensure the functionality remains same if the flag isn't set.

# After Submitting

Update docs with the new `--stream` flag for the `http get` command.

### CODE REVIEW

**Comment by WindSoilder** [comment]
Thank you for the pr!  Would you please share an example about the output when connecting to websocket with `--stream`?  And what if users fire `http get --stream` with not websocket endpoing?

**Comment by alex-kattathra-johnson** [comment]
It would just return a regular http response, what would you recommend, @WindSoilder ?

**Review by WindSoilder — CHANGES_REQUESTED** [review]
Sorry for lately reply.  I left some comments

**Comment by alex-kattathra-johnson** [comment]
Rebased the branched to get a linear commit graph. Gonna make the changes you suggested, @WindSoilder

File: crates/nu-command/src/network/http/client.rs [hunk 1]
```diff
@@ -43,6 +48,128 @@ impl From<Option<ContentType>> for BodyType {
...
+        if let Some(timeout) = timeout {
+            cr.deadline = Some(Instant::now() + timeout);
+        }
+        cr
+    }
+}
+
+impl Read for ChannelReader {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        let rx = self.rx.clone();
```
[the remark this answers is not in the document]
Comment (alex-kattathra-johnson) [reply crates/nu-command/src/network/http/client.rs:71]: I think, because the `rx` is an `Arc` whose reference count needs to be updated for each read.

**Comment by WindSoilder** [comment]
Thanks!  The implementation looks good to me, but sadly I found tungstenite-rs doesn't support wasm: https://github.com/snapview/tungstenite-rs/issues/417
So there are two ways possible:
1. find another websocket crate which supports wasm, e.g: wasm-bindgen
2. guard the `--stream` flag under `os` feature.

**Comment by alex-kattathra-johnson** [comment]
@WindSoilder : Ah! I'm good with option #2. What do you think?

**Comment by cablehead** [comment]
My understanding is that the `http` builtins already stream their body: just checking, this flag / feature is really a websocket upgrade, vs enabling streaming?

**Comment by alex-kattathra-johnson** [comment]
@cablehead: That is correct, it's possible the `--stream` isn't the best name for the flag. The ureq http client currently attempts to unchunk http responses from endpoints that might not have an explicit EOF, which causes the command to hand indefinitely without any outputs. This was an attempt to get around that for endpoints that are upgradable to websocket.

**Comment by cablehead** [comment]
> which causes the command to hang indefinitely without any outputs

I wonder if you're seeing this general issue Nushell has with blocking IO: https://github.com/nushell/nushell/issues/12129 ? not sure if it's possible to confirm? 

> it's possible the --stream isn't the best name for the flag

I wonder if it warrants a whole new subcommand, or perhaps it could be driven from the url protocol? e.g. wss://

**Comment by cablehead** [comment]
Just passing along my takeaway from last night’s open mic: my understanding is the consensus is that advanced HTTP features should be handled through a plugin. The idea is that the built-in HTTP features stay minimal—just enough to support tasks like a plugin that provides WebSocket support—while being easier to maintain and secure. Seeing this as a plugin seems like a great way to make these advanced features happen!?

**Comment by alex-kattathra-johnson** [comment]
@cablehead: That's a good point, I guess I'm somewhat of a noob to the nushell ecosystem and haven't played around with plugins much. I think a plugin will suffice. I could even just create a plugin for my own personal use and contribute it upstream if others might find it useful.

**Comment by fdncred** [comment]
Lots of people do that @alex-kattathra-johnson. Take a look at our [awesome-nu repo](https://github.com/nushell/awesome-nu) and see.

**Pull request closed without merging**